File: cpswAsynPortDriverApp/deviceLibrarySrc/AxiAd5780.cc
#include "AxiAd5780.h"

#define addrSize 4

AxiAd5780 IAxiAd5780::create(const char *name)
{
AxiAd5780Impl v = CShObj::create<AxiAd5780Impl>(name);
Field f;
        f = IIntField::create("dacRefreshRate", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x10*addrSize);
        f = IIntField::create("dacData", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x30*addrSize);
        f = IIntField::create("debugMux", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x80*addrSize);
        f = IIntField::create("debugData", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x90*addrSize);
        f = IIntField::create("sdoDisable", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0xA0*addrSize);
        f = IIntField::create("binaryOffset", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0xA1*addrSize);
        f = IIntField::create("dacTriState", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0xA2*addrSize);
        f = IIntField::create("opGnd", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0xA3*addrSize);
        f = IIntField::create("rbuf", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0xA4*addrSize);
        f = IIntField::create("halfSckPeriod", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0xA5*addrSize);
        f = IIntField::create("dacRst", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0xFE*addrSize);
        return v;
}

CAxiAd5780Impl::CAxiAd5780Impl(Key &key, const char *name) : CMMIODevImpl(key, name, 0x00010000 >> 2, LE)
{
}

File: util/templates/AxiAd5780.py
from __future__ import print_function
import os
from Cheetah.Template import Template

d = { 'name': "AxiAd5780",
     'registers': [{ 'name'   : "dacRefreshRate",
                    'address' : "0x10",
                    'type'    : "Status" },
                   {'name'    : "dacData",
                    'address' : "0x30",
                    'type'    : "Status" },
                   {'name'    : "debugMux",
                    'address' : "0x80",
                    'type'    : "Configuration" },
                   {'name'    : "debugData",
                    'address' : "0x90",
                    'type'    : "Configuration" },
                   {'name'    : "sdoDisable",
                    'address' : "0xA0",
                    'type'    : "Configuration" },
                   {'name'    : "binaryOffset",
                    'address' : "0xA1",
                    'type'    : "Configuration" },
                   {'name'    : "dacTriState",
                    'address' : "0xA2",
                    'type'    : "Configuration" },
                   {'name'    : "opGnd",
                    'address' : "0xA3",
                    'type'    : "Conifiguration" },
                   {'name'    : "rbuf",
                    'address' : "0xA4",
                    'type'    : "Configuration" },
                   {'name'    : "halfSckPeriod",
                    'address' : "0xA5",
                    'type'    : "Status" },
                   {'name'    : "dacRst",
                    'address' : "0xFE",
                    'type'    : "Configuration" } ],
     'commands': [{ 'name'   : "dacRst"} ]
   }



os.mkdir('AxiAd5780')
os.mkdir('AxiAd5780/deviceLibrary')
os.mkdir('AxiAd5780/AxiAd5780Src')
os.mkdir('AxiAd5780/Db')
os.mkdir('AxiAd5780/ui')

#device library
f=open('AxiAd5780/deviceLibrary/AxiAd5780.cc', 'w')
print(Template( file='builder.cc.tmpl', searchList = [d] ), file=f)
f.close()

f=open('AxiAd5780/deviceLibrary/AxiAd5780.h', 'w')
print(Template( file='builder.h.tmpl', searchList = [d] ), file=f)
f.close()

#asyn driver
f=open('AxiAd5780/AxiAd5780Src/Makefile', 'w')
print(Template( file='Makefile.tmpl', searchList = [d] ), file=f)
f.close()

f=open('AxiAd5780/AxiAd5780Src/AxiAd5780Driver.cpp', 'w')
print(Template( file='templateDriver.cpp.tmpl', searchList = [d] ), file=f)
f.close()

f=open('AxiAd5780/AxiAd5780Src/AxiAd5780Driver.h', 'w')
print(Template( file='templateDriver.h.tmpl', searchList = [d] ), file=f)
f.close()

f=open('AxiAd5780/AxiAd5780Src/AxiAd5780DriverSupportInclude.dbd', 'w')
print(Template( file='templateDriverSupportInclude.dbd.tmpl', searchList = [d] ), file=f)
f.close()

#database substitutions
f=open('AxiAd5780/Db/AxiAd5780.substitutions', 'w')
print(Template( file='template.substitutions.tmpl', searchList = [d] ), file=f)
f.close()

#gui
f=open('AxiAd5780/ui/AxiAd5780.ui', 'w')
print(Template( file='templateGui.ui.tmpl', searchList = [d] ), file=f)
f.close()
